fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/MetaDataEvolutionValidator.java

```diff
@@ -0,0 +1,906 @@
+            FieldDescriptor oldField = oldFieldEntry.getValue();
+            FieldDescriptor newField = newFieldMap.get(fieldNumber);
+            if (newField != null) {
+                validateField(oldField, newField, seenDescriptors);
+            } else {
+                throw new MetaDataException("field removed from message descriptor", LogMessageKeys.FIELD_NAME, oldField.getName());
+            }
+        }
+        // Validate that none of the new fields are "required", which would make reading back any existing record impossible
+        for (Map.Entry<Integer, FieldDescriptor> newNumberToDescriptorEntry : newFieldMap.entrySet()) {
+            if (!oldFieldMap.containsKey(newNumberToDescriptorEntry.getKey())) {
+                final FieldDescriptor newFieldDescriptor = newNumberToDescriptorEntry.getValue();
+                if (newFieldDescriptor.isRequired()) {
+                    throw new MetaDataException("required field added to record type", LogMessageKeys.FIELD_NAME, newFieldDescriptor.getName());
+                }
+            }
+        }
+    }
+
+    @Nonnull
+    private Map<Integer, FieldDescriptor> getFieldNumberToDescriptorMap(@Nonnull Descriptor descriptor) {
+        final List<FieldDescriptor> fieldDescriptors = descriptor.getFields();
+        if (fieldDescriptors.isEmpty()) {
+            return Collections.emptyMap();
+        }
+        final Map<Integer, FieldDescriptor> numberToDescriptorMap = new HashMap<>(fieldDescriptors.size() * 2);
+        for (FieldDescriptor fieldDescriptor : fieldDescriptors) {
+            numberToDescriptorMap.put(fieldDescriptor.getNumber(), fieldDescriptor);
+        }
+        return numberToDescriptorMap;
+    }
+
+    private void validateField(@Nonnull FieldDescriptor oldFieldDescriptor, @Nonnull FieldDescriptor newFieldDescriptor,
+                               @Nonnull Set<Pair<Descriptor, Descriptor>> seenDescriptors) {
+        if (!oldFieldDescriptor.getName().equals(newFieldDescriptor.getName())) {
+            // TODO: Field renaming should be allowed with some caveats about if the field is indexed or not
+            throw new MetaDataException("field renamed",
+                    LogMessageKeys.OLD_FIELD_NAME, oldFieldDescriptor.getName(),
+                    LogMessageKeys.NEW_FIELD_NAME, newFieldDescriptor.getName());
+        }
+        if (!oldFieldDescriptor.getType().equals(newFieldDescriptor.getType())) {
+            validateTypeChange(oldFieldDescriptor, newFieldDescriptor);
+        }
+        if (oldFieldDescriptor.isRequired() && !newFieldDescriptor.isRequired()) {
+            throw new MetaDataException("required field is no longer required",
+                    LogMessageKeys.FIELD_NAME, oldFieldDescriptor.getName());
+        } else if (oldFieldDescriptor.isOptional() && !newFieldDescriptor.isOptional()) {
+            // TODO: In theory, optional -> repeated is okay, but only if the field is not indexed
```

Comment:
Or if the index definition is changed to fan-out at the same time, I think. Not for this branch, but maybe update the comment?

Reply:
That sounds like more than this branch attempts.